### PR TITLE
Fixed strinject issue on mac

### PR DESCRIPTION
**Pull request by jjr840430**

### CODE REVIEW

**Comment by jjr840430** [comment]
@erikbor It would be work on Mac without any issues. :)
@drystone Thanks for your help to solve the problems.

File: examples/stringinject/strinject_test.c [hunk 1]
```diff
@@ -60,7 +60,7 @@ static void test_fread(void)
 		if (!fp)
```
Comment (drystone) [inline examples/stringinject/strinject_test.c]: line 57: `size_t len;` to avoid confusion.

File: src/file.c [hunk 2]
```diff
@@ -531,18 +536,31 @@ size_t RETRACE_IMPLEMENTATION(fwrite)(const void *ptr, size_t size, size_t nmemb
 	event_info.return_value_type = PARAMETER_TYPE_INT;
 	event_info.return_value = &r;
 	event_info.logging_level = RTR_LOG_LEVEL_NOR;
-	retrace_log_and_redirect_before(&event_info);
 
-	if (rtr_str_inject(STRINJECT_FUNC_FWRITE, (void *) ptr, size * nmemb)) {
+	if (size == 1 && rtr_str_inject(STRINJECT_FUNC_FWRITE, (void *) ptr, size * nmemb, &inject_buffer, &inject_len)) {
```
Comment (drystone) [inline src/file.c]: `size == 1` so `size * nmemb == nmemb`

Comment (jjr840430) [reply src/file.c]: yes, right

Comment (drystone) [reply src/file.c]: also `fwrite("hello", 5, 1, f)` is probably the same as `fwrite("hello", 1, 5, f)` but only one will get injected. Is that intentional?

Comment (jjr840430) [reply src/file.c]: yes, not possible to inject using fwrite when size is more than 1
but other functions are possible

File: src/file.c [hunk 3]
```diff
@@ -531,18 +536,31 @@ size_t RETRACE_IMPLEMENTATION(fwrite)(const void *ptr, size_t size, size_t nmemb
...
-	r = real_fwrite(ptr, size, nmemb, stream);
-	if (r < size * nmemb)
+	retrace_log_and_redirect_before(&event_info);
+
+	r = real_fwrite(enable_inject ? inject_buffer : ptr,
+			size,
+			enable_inject ? inject_len : nmemb,
+			stream);
+
+	if (r < (inject_len > 0 ? size * nmemb : size * inject_len))
```
Comment (drystone) [inline src/file.c]: This looks odd to me - equates to `if (r < 0)` when `inject_len == 0` ?

Comment (jjr840430) [reply src/file.c]: inject_len is always more than 0

Comment (jjr840430) [reply src/file.c]: ah, maybe it's wrong, thanks for your finding issue

Comment (jjr840430) [reply src/file.c]: should be inject_len > size * nmemb

**Comment by jjr840430** [comment]
@drystone ah, I forgot to replace when the type is SINGLE_HEX. I will modify.
But it needs to insert strings for other types.

**Comment by drystone** [comment]
@jjr840430 I don't think it's sensible for a write to write more data than intended - you're messing with the universe :)

**Comment by jjr840430** [comment]
@drystone I don't want to mess with the universe, but it's @erikbor 's request. :)

**Comment by ronaldtse** [comment]
Let's get @erikbor to confirm if he wants to mess with the universe 😉

**Comment by erikbor** [comment]
@jjr840430 works really well! I'm going to fuzz loads of software this week :) Thanks for this really good implementation!!!

**Comment by erikbor** [comment]
@jjr840430 yes we do want to mess with the universe! If we send longer strings than intended using a retraced client then we could potentially cause a buffer overflow in a server/daemon :D

**Comment by erikbor** [comment]
@jjr840430 I also cannot get the INJECT_SINGLE_HEX to work:

```sh
$ cat ~/tmp/send.conf
stringinject,INJECT_SINGLE_HEX,send,3:RANDOM,1
$ ./retrace -f ~/tmp/send.conf ~/tmp/send
<SNIP>
(2837) memcpy(0x7fff4e0008a0, 0x7fff4bc00097, 7, ) = 0x7fff4e0008a0
(2837) sendto(0 [socket[172.217.24.36:80]], 0x7fff5c7dc300, 12, 0, 0x0, 0, ) = 12
	0000000	 7465 7374 7465 7374 7465 7374                     | testtesttest
(2837) send(0 [socket[172.217.24.36:80]], 0x7fff5c7dc300, 12, ) = 12
	0000000	 7465 7374 7465 7374 7465 7374                     | testtesttest
(2837) memcpy(0x7fff4d001e00, 0x103423fa6, 5, ) = 0x7fff4d001e00
sent
<SNIP>
$ ./retrace -f ~/tmp/send.conf ~/tmp/send
(2893) sendto(0 [socket[172.217.24.36:80]], 0x7fff5e7ef300, 12, 0, 0x0, 0, ) = 12
	0000000	 7465 7374 7465 7374 7465 7374                     | testtesttest
(2893) send(0 [socket[172.217.24.36:80]], 0x7fff5e7ef300, 12, ) = 12
	0000000	 7465 7374 7465 7374 7465 7374                     | testtesttest
(2893) memcpy(0x7fff4d800a00, 0x101410fa6, 5, ) = 0x7fff4d800a00
sent
(2893) free(0x0, )
<SNIP>
```

**Comment by jjr840430** [comment]
@erikbor Can't you get working of stringinject?
Please refer the following link for configuration settings.

https://github.com/riboseinc/retrace/tree/master/examples/stringinject

**Comment by erikbor** [comment]
@jjr840430 I got it working now, it was indeed some configuration syntax problem. So far it works really well. The only thing that is missing is the "mess with the universe" buffer overflow sending. Once that's in then we can start fuzzing a lot of daemons :) Cheers

**Comment by ronaldtse** [comment]
Since it's working I'm going to first merge this. @erikbor can you create a new issue on messing with the universe? Thanks!

**Pull request merged**